### ISSUE TITLE
apk: scripts tar file a lot bigger than it needs to be

### ISSUE THREAD

**Issue by efahl** [openwrt/openwrt#17108]
### Describe the bug

I just noticed this one and thought it would be good to log it here before I forget.

The `scripts.tar` file is mostly null padding.  Can we gzip it?  Or somehow get tar to reduce block size or ???  Seems like an easy 250k (or more) savings...

```
# ll /lib/apk/db/scripts.tar
-rw-r--r--    1 root     root        311296 Nov 27 00:42 /lib/apk/db/scripts.tar

# tar tvf /lib/apk/db/scripts.tar | awk '{x += $3} END {print x" bytes"}'
50450 bytes
```

### OpenWrt version

r28200-c857145e03

### OpenWrt release

SNAPSHOT

### OpenWrt target/subtarget

x86/64

### Device

generic

### Image kind

Official downloaded image

### Terms

- [X] I am reporting an issue for OpenWrt, not an unsupported fork.

**Comment by robimarko** [comment]
I guess we could make APK compress it since its already supports gzip AFAIK

**Comment by efahl** [comment]
Just saw this again when reviewing some other things, and somehow it appears to be even worse than before.  It's not so bad when it's in squashfs `/rom`, but the first time you add or remove a package, it goes into `/overlay` and explodes...

```
$ cat /lib/apk/db/scripts.tar | wc -c
624640

$ gzip -c /lib/apk/db/scripts.tar | wc -c
24973
```

**Comment by robimarko** [comment]
The issue is that scripts.tar is generated during build and installed in the squashfs, and then when you add or remove a package it will regenerate it and therefore you end up with quite a large diff in OverlayFS so it will be bigger than the original scripts.tar itself.

I remember looking into this and compressing it quite a while ago but I dont remember whether that helped at all.

**Comment by efahl** [comment]
Yeah, I'm anticipating someone with a marginal device looking at `df -h`, seeing half a meg free, then trying to install a 10kb package and running out of space.  "Well, it's like this..."

I've dug through apk-tools/src/database.c and I'm sure that I don't want to deal with that.  Maybe I should open an issue over there?

**Comment by robimarko** [comment]
Yeah, please open an issue there, there is likely some workaround

**Comment by efahl** [comment]
https://gitlab.alpinelinux.org/alpine/apk-tools/-/issues/11154

**Comment by efahl** [comment]
@fabled got us an implementation in record time (👍🏼), so I've been doing preliminary testing on both host and x86 device.  (Preliminary as in: I've only applied the specific changes as a patch without any other changes since our last version bump to apk-tools-3.0.0_rc6; everything is working as expected.)

https://gitlab.alpinelinux.org/alpine/apk-tools/-/commit/012cdcfdf981453b572a70c9568e8df6d0fe0ca1

@robimarko 
Just one question for you:
Should we wait for rc9 and then grab that, or should I update apk-tools now to this commit?

This commit missed the apk-tools rc8 tag by a few days, so looking at their cadence, might be a month or two (or a lot less, it all depends) for the rc9...

Subsequent to this making it into the code, changes to our build are pretty trivial, I've got it working with a one-liner at
https://github.com/openwrt/openwrt/blob/971c058cf55f85a295378031ba7f7435cb2a0f6a/include/rootfs.mk#L94
Simply zip scripts.tar in the rootfs after we've finished creating it:
```
                done; \
-               if [ -z "$(CONFIG_USE_APK)" ]; then \
+               if [ -n "$(CONFIG_USE_APK)" ]; then \
+                       gzip -9 ./lib/apk/db/scripts.tar; \
+               else \
                        $(if $(IB),,awk -i inplace \
```

**Comment by efahl** [comment]
Device tests:
```
$ ll /lib/apk/db/scripts.t*
-rw-r--r--    1 root     root         23113 Nov 14 10:11 /lib/apk/db/scripts.tar.gz

$ tar tzvf /lib/apk/db/scripts.tar.gz | wc
      609      3654     77813

$ apk del nmap
OK: 64 MiB in 298 packages

$ tar tzvf /lib/apk/db/scripts.tar.gz | wc
      609      3654     77813

$ apk add nmap
(1/1) Installing nmap (7.95-r1)
nmap-7.95-r1.post-install: Executing script...
OK: 76 MiB in 299 packages

$ tar tzvf /lib/apk/db/scripts.tar.gz | wc
      612      3672     78156

$ tar tzvf /lib/apk/db/scripts.tar.gz | grep nmap
-rwxr-xr-x 0/0       226 1969-12-31 16:00:00 nmap-7.95-r1.X16693cd5c3df1f5a54e286e77a9821da68ec2b555.post-install
-rwxr-xr-x 0/0       164 1969-12-31 16:00:00 nmap-7.95-r1.X16693cd5c3df1f5a54e286e77a9821da68ec2b555.pre-deinstall
-rwxr-xr-x 0/0       257 1969-12-31 16:00:00 nmap-7.95-r1.X16693cd5c3df1f5a54e286e77a9821da68ec2b555.post-upgrade

$ apk del nmap
(1/1) Purging nmap (7.95-r1)
nmap-7.95-r1.pre-deinstall: Executing script...
OK: 64 MiB in 298 packages

$ tar tzvf /lib/apk/db/scripts.tar.gz | wc
      609      3654     77813

$ tar tzvf /lib/apk/db/scripts.tar.gz | grep nmap

$ apk add nmap
(1/1) Installing nmap (7.95-r1)
nmap-7.95-r1.post-install: Executing script...
OK: 76 MiB in 299 packages

$ tar tzvf /lib/apk/db/scripts.tar.gz | grep nmap
-rwxr-xr-x 0/0       226 1969-12-31 16:00:00 nmap-7.95-r1.X16693cd5c3df1f5a54e286e77a9821da68ec2b555.post-install
-rwxr-xr-x 0/0       164 1969-12-31 16:00:00 nmap-7.95-r1.X16693cd5c3df1f5a54e286e77a9821da68ec2b555.pre-deinstall
-rwxr-xr-x 0/0       257 1969-12-31 16:00:00 nmap-7.95-r1.X16693cd5c3df1f5a54e286e77a9821da68ec2b555.post-upgrade

$ tar tzvf /lib/apk/db/scripts.tar.gz | wc
      612      3672     78156

$ ll /lib/apk/db/scripts.t*
-rw-r--r--    1 root     root         23205 Nov 14 10:12 /lib/apk/db/scripts.tar.gz
```

**Comment by robimarko** [comment]
@efahl Why not simply backport this change and get this into OpenWrt ASAP as this is a problem for the next release as well

**Comment by efahl** [comment]
Will do, PR incoming...

**Comment by efahl** [comment]
> this is a problem for the next release as well

Ha, I forgot to hit "Comment" yesterday, when I was going to say: Yes, my major concern is that 25.x would sprint ahead and this would get missed, then I'd have to explain why removing a package caused "disk full"(!) to someone.
